pkg/health/probe/probe.go

```diff
@@ -18,78 +18,135 @@ package probe
+// new processes all the time when the probe is invoked.
+func NewHealthProbe(ctx context.Context, options ...Option) (*HealthProbe, error) {
+
 	opts := &probeOptions{
-		cmd:         "grpc_health_probe",
-		addr:        fmt.Sprintf("-addr=%v", health.DefaultURL),
-		service:     "-service=",
-		rpcTimeout:  "-rpc-timeout=400ms",
-		connTimeout: "-connect-timeout=400ms",
+		rpcTimeout:  time.Second,
+		connTimeout: time.Second,
+		userAgent:   "meridio-grpc-health-probe",
 	}
+
 	for _, opt := range options {
 		opt(opts)
 	}
-	_, err := exec.LookPath(opts.cmd)
-	if err != nil {
-		return nil, errors.New("GrpcHealthProbe not found, err: " + err.Error())
-	}
 
-	return &GrpcHealthProbe{
-		cmd:         opts.cmd,
+	hp := &HealthProbe{
+		userAgent:   opts.userAgent,
 		addr:        opts.addr,
 		service:     opts.service,
-		rpcTimeout:  opts.rpcTimeout,
 		connTimeout: opts.connTimeout,
+		rpcTimeout:  opts.rpcTimeout,
 		spiffe:      opts.spiffe,
-	}, nil
+	}
+
+	if hp.spiffe {
+		// fetch X509Source once to avoid recurring load on spire
+		source, err := workloadapi.NewX509Source(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("error getting x509 source, %w", err)
+		}
+		svid, err := source.GetX509SVID()
+		if err != nil {
+			return nil, fmt.Errorf("error getting x509 svid, %w", err)
+		}
+		log.FromContextOrGlobal(ctx).V(1).Info("GetX509Source for health probe", "sVID", svid.ID)
+		hp.source = source
+	}
+
+	return hp, nil
 }
 
 // String -
-func (ghp *GrpcHealthProbe) String() string {
-	return fmt.Sprintf("%v %v %v %v %v %v", ghp.cmd, ghp.addr, ghp.service, ghp.connTimeout, ghp.rpcTimeout, ghp.spiffe)
+func (hp *HealthProbe) String() string {
+	return fmt.Sprintf("addr=%v, service=%v, userAgent=%v, connTimeout=%v, rpcTimeout=%v, spiffe=%v",
+		hp.addr, hp.service, hp.userAgent, hp.connTimeout.String(), hp.rpcTimeout.String(), hp.spiffe)
 }
 
-// Run -
-// Runs gRPC health checking probe
-// Returns nil if status is HealthCheckResponse_SERVING, and err otherwise
-func (ghp *GrpcHealthProbe) Run(ctx context.Context) error {
-	stdoutStderr, err := exec.CommandContext(ctx, ghp.cmd, ghp.addr, ghp.service,
-		ghp.connTimeout, ghp.rpcTimeout, ghp.spiffe).CombinedOutput()
+// Request -
+// Sends query to check health of gRPC services exposing their status through
+// gRPC Health Checking Protocol.
+//
+// Returns no error if queried service is SERVING.
+func (hp *HealthProbe) Request(ctx context.Context) error {
+	opts := []grpc.DialOption{
+		grpc.WithUserAgent(hp.userAgent),
+		grpc.WithBlock(),
+	}
+
+	if hp.spiffe {
+		opts = append(opts, grpc.WithTransportCredentials(
+			credentials.NewTLS(tlsconfig.MTLSClientConfig(hp.source, hp.source, tlsconfig.AuthorizeAny())),
+		))
+	} else {
+		opts = append(opts, grpc.WithTransportCredentials(
+			insecure.NewCredentials(),
+		))
+	}
+
+	dialCtx, dialCancel := context.WithTimeout(ctx, hp.connTimeout)
+	defer dialCancel()
+	conn, err := grpc.DialContext(dialCtx, hp.addr, opts...)
```

Comment:
Why not Dial and create the connection as part of the NewHealthProbe func?

Reply:
Mostly I did not want to maintain a connection for the sake of seldom short msg exchange.

Also, according to this [comment](https://github.com/networkservicemesh/deployments-k8s/issues/6779#issuecomment-1191316798) there might issues upon cert expirations. I haven't dig into the details, however did not want to riks any problems.